src/sources/docker.rs

```diff
@@ -0,0 +1,802 @@
+    last_log: Option<(DateTime<FixedOffset>, u64)>,
+    /// generation of ContainerState at event_stream creation
+    generation: u64,
+}
+
+/// Kept by main to keep track of container state
+struct ContainerState {
+    /// None if there is a event_stream of this container.
+    info: Option<ContainerLogInfo>,
+    /// True if Container is currently running
+    running: bool,
+    /// Of running
+    generation: u64,
+}
+
+impl ContainerState {
+    /// Container docker ID
+    /// Unix timestamp of event which created this struct
+    fn new(id: String, created: i64) -> Self {
+        let info = ContainerLogInfo {
+            id: id,
+            created,
+            last_log: None,
+            generation: 0,
+        };
+        ContainerState {
+            info: Some(info),
+            running: true,
+            generation: 0,
+        }
+    }
+
+    /// If info is present, runs event stream
+    fn run_event_stream(
+        &mut self,
+        out: &Sender<Event>,
+        main: &Sender<ContainerLogInfo>,
+        docker: &Docker,
+    ) {
+        if let Some(info) = self.info.take() {
+            run_event_stream(self, info, out.clone(), main.clone(), docker);
+        }
+    }
+}
+
+/// Returns main future which listens for events coming from docker, and maintains
+/// a fan of event_stream futures.
+/// Where each event_stream corresponds to a runing container marked with ContainerLogInfo.
+/// While running, event_stream streams Events to out channel.
+/// Once a log stream has ended, it sends ContainerLogInfo back to main.
+///
+/// Future  channel     Future      channel
+///           |<---- event_stream ---->out
+/// main <----|<---- event_stream ---->out
+///           | ...                 ...out
+///
+#[allow(dead_code)]
+fn docker_source(
+    config: DockerConfig,
+    out: Sender<Event>,
+) -> Result<impl Future<Item = (), Error = ()>, String> {
+    // TODO: async_docker should be replaced with bollard once it supports events
+    // ?NOTE: Requiers sudo privileges, or docker group membership.
+    // Without extra configuration of docker on user side, there is no way around above.
+    let docker_for_events = async_docker::new_docker(None).map_err(|error| {
+        error!(message="Error connecting to docker server",%error);
+        "Failed to connect to docker server".to_owned()
+    })?;
+
+    // ?NOTE: Requiers sudo privileges, or docker group membership.
+    // Without extra configuration of docker on user side, there is no way around above.
+    let docker = Docker::connect_with_local_defaults().map_err(|error| {
+        error!(message="Error connecting to docker server",%error);
+        "Failed to connect to docker server".to_owned()
+    })?;
+
+    // Channel could be unbounded, since there won't be millions of containers,
+    // but bounded should be more performant in most cases.
+    let (main_send, mut main_recv) = mpsc::channel::<ContainerLogInfo>(100);
```

Comment:
Bounded implementation should have one or two memory/atomic accesses less. 
But considering the amount gained and times it will occur, this is highly unnecessary, so I'll revert this back to unbounded version.

Reply:
Bummer